lib/src/package_config_impl.dart

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import "package_config.dart";
+export "package_config.dart";
+import "util.dart";
+
+class SimplePackageConfig implements PackageConfig {
+  static const int _maxVersion = 2;
+
+  final int version;
+  final Map<String, Package> _packages;
+
+  SimplePackageConfig(int version, Iterable<Package> packages)
+      : version = _validateVersion(version),
+        _packages = _validatePackages(packages);
+
+  SimplePackageConfig._(int version, Iterable<SimplePackage> packages)
+      : version = _validateVersion(version),
+        _packages = {for (var package in packages) package.name: package};
+
+  /// Creates empty configuration.
+  ///
+  /// The empty configuration can be used in cases where no configuration is
+  /// found, but code expects a non-null configuration.
+  const SimplePackageConfig.empty()
+      : version = 1,
+        _packages = const <String, Package>{};
+
+  static int _validateVersion(int version) {
+    RangeError.checkValueInInterval(version, 1, _maxVersion, "version");
+    return version;
+  }
+
+  static Map<String, Package> _validatePackages(Iterable<Package> packages) {
+    Map<String, Package> result = {};
+    for (var package in packages) {
+      if (package is! _SimplePackage) {
+        // _SimplePackage validates these properties.
+        try {
+          _validatePackageData(package.name, package.root,
+              package.packageUriRoot, package.languageVersion);
+        } catch (e) {
+          throw ArgumentError.value(
+              packages, "packages", "Package ${package.name}: ${e.message}");
+        }
+      }
+      var name = package.name;
+      if (result.containsKey(name)) {
+        throw ArgumentError.value(name, "packages", "Duplicate package name");
+      }
+      result[name] = package;
+    }
+
+    // Check that no root URI is a prefix of another.
```

Comment:
Do they depend on the root package's root directory or just its lib directory. The latter is supported.

I would prefer to not have lookup need to go through all packages to find a match instead of finding just the first one.
Maybe it's not a problem, I'll see how bad it might be - maybe I can recognize the situation and link the packages somehow.

Reply:
Flutter packages violate the original package conventions in at least two important ways:

- They share assets that aren't under `lib` directories (typically they are in a top level `assets` directory, and become available to any other package effectively).
- They explicitly endorse putting entire packages under `example` because its a lot more useful to do that. They are fully encapsulated examples that can be copy/pasted and used as starting points etc.

Effectively, they are relying on the fact that `pub` downloads the entire package and not just the `lib` dir. Pub will also have to continue setting the `root` to the actual root of the package and the `packageUriRoot` to `lib/`.

Note that Pub would have to do this anyways because of package executables - pub itself runs code from the `bin` directory of packages.

Reply:
I'll allow nested packages in a single package configuration.

I'm not sure it's required here. If the nested package has its own `package_config.json` (if you run `pub get` in the examples dir, it will), then there won't be a *single* configuration with nested packages. 
Still, it's reasonable to allow it and shouldn't be a problem.